cpp/src_prims/common/buffer_base.hpp

```diff
@@ -58,10 +78,49 @@ class buffer_base {
 
   const_iterator end() const { return _data + _size; }
 
+  void reserve(const size_type new_capacity, cudaStream_t stream) {
+    set_stream(stream);
+    if (new_capacity > _capacity) {
+      value_type* new_data = static_cast<value_type*>(
+        _allocator->allocate(new_capacity * sizeof(value_type), _stream));
+      if (_size > 0) {
+        CUDA_CHECK(cudaMemcpyAsync(new_data, _data, _size * sizeof(value_type),
+                                   cudaMemcpyDefault, _stream));
+      }
+      if (nullptr != _data) {
+        _allocator->deallocate(_data, _capacity * sizeof(value_type), _stream);
+      }
+      _data = new_data;
+      _capacity = new_capacity;
+    }
+  }
```

Comment:
General CUDA memory question: Is it at all possible to shrink the array and let the underlying allocator know that the extra memory is available, since we have more control over that memory manager (RMM)?

Reply:
Interesting idea. I am not sure if RMM supports that. Anyway if I am not missing something it is not possible with the current allocator design in cuML. So I would say out of scope for this PR in any case. Do you agree?

Reply:
It's most definitely out of scope for this PR. It's a question I've been pondering myself and found this to be a good opportunity to ask.